stackslib/src/chainstate/nakamoto/miner.rs

```diff
@@ -0,0 +1,812 @@
+        consensus_hash: &ConsensusHash,
+        // total BTC burn so far
+        total_burn: u64,
+        // VRF proof, if we're starting a _new_ tenure (instead of continuing an existing one)
+        vrf_proof_opt: Option<VRFProof>,
+    ) -> Result<NakamotoBlockBuilder, Error> {
+        let builder = if let Some(parent_nakamoto_header) =
+            parent_stacks_header.anchored_header.as_stacks_nakamoto()
+        {
+            // building atop a nakamoto block
+            // new tenure?
+            if let Some(vrf_proof) = vrf_proof_opt.as_ref() {
+                NakamotoBlockBuilder::new_tenure_from_nakamoto_parent(
+                    parent_tenure_id,
+                    parent_nakamoto_header,
+                    consensus_hash,
+                    total_burn,
+                    vrf_proof,
+                )
+            } else {
+                NakamotoBlockBuilder::continue_tenure_from_nakamoto_parent(
+                    parent_nakamoto_header,
+                    consensus_hash,
+                    total_burn,
+                )
+            }
+        } else if let Some(parent_epoch2_header) =
+            parent_stacks_header.anchored_header.as_stacks_epoch2()
+        {
+            // building atop a stacks 2.x block.
+            // we are necessarily starting a new tenure
+            if let Some(vrf_proof) = vrf_proof_opt.as_ref() {
+                NakamotoBlockBuilder::new_tenure_from_epoch2_parent(
+                    parent_epoch2_header,
+                    &parent_stacks_header.consensus_hash,
+                    consensus_hash,
+                    total_burn,
+                    vrf_proof,
+                )
+            } else {
+                // not allowed
+                warn!("Failed to start a Nakamoto tenure atop a Stacks 2.x block -- missing a VRF proof");
+                return Err(Error::ExpectedTenureChange);
+            }
+        } else {
+            // not reachable -- no other choices
+            return Err(Error::InvalidStacksBlock(
+                "Parent is neither a Nakamoto block nor a Stacks 2.x block".into(),
+            ));
+        };
+
+        Ok(builder)
+    }
+
+    /// This function should be called before `tenure_begin`.
+    /// It creates a MinerTenureInfo struct which owns connections to the chainstate and sortition
+    /// DBs, so that block-processing is guaranteed to terminate before the lives of these handles
+    /// expire.
+    pub fn load_tenure_info<'a>(
+        &self,
+        chainstate: &'a mut StacksChainState,
+        burn_dbconn: &'a SortitionDBConn,
+        tenure_start: bool,
+    ) -> Result<MinerTenureInfo<'a>, Error> {
+        debug!("Nakamoto miner tenure begin");
+
+        let burn_tip = SortitionDB::get_canonical_chain_tip_bhh(burn_dbconn.conn())?;
+        let burn_tip_height =
+            SortitionDB::get_canonical_burn_chain_tip(burn_dbconn.conn())?.block_height as u32;
```

Comment:
`u64 as u32` can truncate. Yes it will take thousands of years at current Bitcoin block rate, but suppose one day Bitcoin decides on faster blocks, or this code gets used elsewhere (like Subnets, or another chain). Or suppose we add integration tests with unrealistic values for `burn_height`. This sort of bug is unlikely to occur, but very difficult to track down, so it's better to make it impossible

```suggestion            
 u32::try_from(SortitionDB::get_canonical_burn_chain_tip(burn_dbconn.conn())?.block_height).expect("block_height overflow");
```